### DISCUSSION TITLE
Is there a pattern established to allow `react-jss` to work in `/app` folder

### DISCUSSION THREAD

**Discussion by mildfuzz** [vercel/next.js#47355] [Help] [answered]
### Summary

So we've been using a variation of this [example](https://github.com/vercel/next.js/blob/canary/examples/with-react-jss/pages/_document.tsx) in order to get `react-jss` working with previous versions of Next, but are looking to upgrade.

Trouble is, this depends on accessing certain contexts and functions which don't seem to be available to the `RootLayout` in the `./app` folder.

Is there a work around here?

**Comment by sscaff1** [comment, marked as answer]
You can create a client side component like this:

```tsx
'use client';

import { useServerInsertedHTML } from 'next/navigation';
import { useEffect, useState } from 'react';
import { JssProvider, SheetsRegistry, createGenerateId } from 'react-jss';

function JSSCacheProvider({ children }) {
  const [{ generateId, registry }] = useState(() => {
    return { registry: new SheetsRegistry(), generateId: createGenerateId() };
  });

  useEffect(() => {
    const style = document.body.querySelector('#server-side-styles');
    if (style && style.parentNode) {
      style.parentNode.removeChild(style);
    }
  }, []);

  useServerInsertedHTML(() => {
    return <style id="server-side-styles">{registry.toString()}</style>;
  });

  return (
    <JssProvider generateId={generateId} registry={registry}>
      {children}
    </JssProvider>
  );
}

export default JSSCacheProvider;
```

Then import this into your app/layout.tsx and wrap your children with it. react-jss works, but you ned up with this error in the console:

app-index.js:35 Warning: Prop `className` did not match. Server: "root-8-2-1" Client: "root-0-2-1"

Before it seems like when client side took over, the server-side stylesheet could safely be removed. Now I'm not sure what's going on. If you remove them all it doesn't work. It also seems to generate two tags (which is why I'm removing the one from the body). 

Note that all components that use react-jss must have be a client component (i.e. using `use client`). It's strange. Even though my component is listed as `use client`, it's still being rendered by the server on a server side request which is very confusing. So it's throwing the warning as a result of that. That's the reason you can't remove all of the server-side stylesheets as well. So this kind of works. Not sure how to make this better. Hope this helps.

**Comment by ilyadoroshin** [comment]
I've recently found a solution to fix the warning. While debugging, I discovered that the className is generated by this line:
<img width="614" alt="image" src="https://github.com/vercel/next.js/assets/3414279/4dfd83a8-0bc4-4b6e-9900-9637d0f4bee1">

I noticed that `moduleId` increments every time I make a change in the code. Upon further inspection of the `moduleId`, I found the increment:
<img width="686" alt="image" src="https://github.com/vercel/next.js/assets/3414279/118054d0-20ea-4d30-82a1-eef16afdc044">

The fix is to provide our own `generateId` function that doesn’t use the `moduleId`, like this:
```tsx
const [{ generateId, registry }] = useState(() => {
  let ruleCounter = 0;

  return {
    registry: new SheetsRegistry(),
    generateId: (rule, sheet) => {
      ruleCounter++;
      const prefix = sheet?.options.classNamePrefix;
      return `${prefix + rule.key}-${ruleCounter}`;
    },
  };
});
```

**Reply by ilyadoroshin** [reply]
To remove all of the ssr-style sheets:
```tsx
  React.useEffect(() => {
    const styles = document.body.querySelectorAll(`#server-side-styles`);
    styles.forEach((style) => {
      if (style && style.parentNode) {
        style.parentNode.removeChild(style);
      }
    });
  }, []);
```
